File: backend/commands/ip.go
package commands

import (
	"backend/types"
	"bytes"
	"encoding/json"
	"fmt"
	"net/http"
	"net/url"
	"strings"
)

type IPInfo struct {
	Status      string  `json:"status"`
	Country     string  `json:"country"`
	CountryCode string  `json:"countryCode"`
	Region      string  `json:"region"`
	RegionName  string  `json:"regionName"`
	City        string  `json:"city"`
	Zip         string  `json:"zip"`
	Lat         float64 `json:"lat"`
	Lon         float64 `json:"lon"`
	Timezone    string  `json:"timezone"`
	Isp         string  `json:"isp"`
	Org         string  `json:"org"`
	As          string  `json:"as"`
	Query       string  `json:"query"`
}

func ip(i *types.Interaction, data *types.ApplicationCommandInteractionData) {
	address, _ := data.GetOption("address")
	resp, err := http.Get("http://ip-api.com/json/" + url.QueryEscape(address.StringValue()) + "&lang=" + i.GuildLocale)
	if err != nil {
		i.Error("Error retrieving info about the IP addrees")
		return
	}

	defer resp.Body.Close()
	var info IPInfo
	if err = json.NewDecoder(resp.Body).Decode(&info); err != nil {
		i.Error("Error retrieving info about the IP address")
		return
	}

	if info.Status == "fail" {
		i.Error("That's not a valid IPv4 or IPv6 address")
		return
	}

	location, err := getGoogleMap(fmt.Sprintf("{%.5f,%.5f}", info.Lat, info.Lon), "", "", 12)
	if err != nil {
		i.Error("Error getting a map of that location")
	}

	i.RespondWithFile(types.InteractionResponse{
		Type: types.ResponseChannelMessageWithSource,
		Data: types.ResponseData{
			Content: fmt.Sprintf("Information about **%s**\n📍 %s, %s, %s :flag_%s:", info.Query, info.City, info.Region, info.Country, strings.ToLower(info.CountryCode)),
			Attachments: []types.Attachment{{
				Filename:    "location.png",
				Description: "The location of that IP address",
				ID:          "0",
			}},
		},
	}, []*bytes.Buffer{bytes.NewBuffer(location)})
}

File: backend/src/commands/ip.ts
import { CommandInteraction, Embed } from "discord-workers";
import { Env } from "..";
import googleMap from "../utils/googleMap";

export const name = "ip";
export const description = "View information about any public IP address";

export function execute(
  int: CommandInteraction,
  env: Env,
  ctx: ExecutionContext
): Promise<Response> {
  const func = async () => {
    const ipAddr = int.getOption<string>("address");
    // prettier-ignore
    const resp = await fetch(`http://ip-api.com/json/${encodeURIComponent(ipAddr!)}&lang=${int.locale}`);
    const body: any = await resp.json();
    if (body.status !== "success")
      return int.send("Couldn't find any information on that IP address.");

    const map = await googleMap(
      env.googleMapsKey,
      `${body.lat},${body.lon}`,
      "hybrid",
      int.locale
    );

    const embed = new Embed()
      .setTitle(body.query)
      .setDescription(`${body.city}, ${body.region}\n${body.country}`)
      .addField("Service Provider", body.isp, true)
      .addField(
        "Local Time",
        new Date().toLocaleString(int.locale, {
          dateStyle: "long",
          timeStyle: "short",
        }),
        true
      )
      .setImage("attachment://map.png");

    return int.send({
      embeds: [embed],
      attachments: [{ name: "map.png", file: map }],
    });
  };
  return Promise.resolve(int.defer());
}
